udata/api/cache.py

```diff
@@ -0,0 +1,118 @@
+
+    Differs from the official Flask Caching Decorator
+    http://flask.pocoo.org/docs/0.12/patterns/viewdecorators/#caching-decorator
+    by dealing with appropriated headers for server AND client-side caches.
+
+    Key points:
+    * Generates the cache key using the full path (using the query string).
+    * Uses `Cache-Control`, `ETag` and `Last-Modified` headers.
+    * Respects browser's hard-refresh.
+
+    Parameters:
+    * `class_instance`: allows the decorator to be applied on a class method.
+    * `check_serverside`: either the client verifies with the server if
+      its cached versions are still valid or not.
+    * `personal`: set to `True` if the response contains sensitive
+      informations only pertinent for a given user.
+    * `client_timeout`: time in seconds before the client considers the
+      cache as being invalid.
+    * `server_timeout`: time in seconds before the server considers the
+      cache as being invalid.
+    * `key_pattern`: you can customize the cache key, it can be either
+      a callable or a string. If it contains `%s`, the placeholder will
+      be replaced by the full path (including query string).
+    * `make_response`: a way to override the default Flask way of
+      creating a response object.
+
+    Inspirations:
+    * http://codereview.stackexchange.com/questions/147038/↩︎
+      improving-the-flask-cache-decorator
+    * https://gist.github.com/glenrobertson/954da3acec84606885f5
+
+    Resources:
+    * https://jakearchibald.com/2016/caching-best-practices/
+    * https://developers.google.com/web/fundamentals/performance/↩︎
+      optimizing-content-efficiency/http-caching
+
+    Image:
+    * https://developers.google.com/web/fundamentals/performance/↩︎
+      optimizing-content-efficiency/images/http-cache-decision-tree.png
+    """
+    def _generate_cache_control():
+        cache_control = ResponseCacheControl()
+        # Before you think the logic is reversed, read Jake's article.
+        if check_serverside:
+            cache_control.no_cache = True
+        else:
+            cache_control.must_revalidate = True
+        if personal:
+            cache_control.private = True
+        else:
+            cache_control.public = True
+        cache_control.max_age = client_timeout
+        return cache_control
+
+    def _generate_cache_key(str_or_callable):
+        if callable(str_or_callable):
+            return str_or_callable()
+        else:
+            try:
+                # Include querystring with `full_path`.
+                return str_or_callable % request.full_path
+            except TypeError:
+                return str_or_callable
+
+    def decorator(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            headers = Headers()
+            headers.add('Cache-Control', _generate_cache_control())
+            cache_key = _generate_cache_key(key_pattern)
+            resp = cache.get(cache_key)
```

Comment:
It don't have to be an user related API: admin and frontend use a session which is using a cookie to authenticate user so with every request, there is a cookie that identify him. And, I just provided multiple different ways to do it in a safer more compatible way.

I'm totally focused on the goal, this why, if the point is to make it incrementally, I dont understand why you prefer to have something which create a coupling between two things that can be totally decoupled HTTP and introduce an already identified issue. And, as you said, it's to handle 2 APIs, so maybe a generic solution with both HTTP and applicative cache is a little bit over-engineered.

I'm so focused I took the time to deeply details my review and my responses.

To me it's like saying: yes we have a performance issue so let's not resolve it but instead let's try to hide it by adding another issue on top of it.

Why not make the HTTP caching part as one step and use the provided and tested `cache.memoize` to cache the data as an another step ? (the signal part of this PR to invalidate cache is still valid with `cache.memoize`) ? If it's not doing the job, yes, we can incrementally raise the complexity to find another solution.

So, to stay focused, what I'm saying is that to solve this performance issue, we have 3 tools that needs to be handled separately:
- HTTP cache headers (which can be solved by this decorator and can benefit all APIs than need  to and can be cached)
- applicative cache, (which is already provided by Flask-Caching and can already be used with `cache.memoize` and for which the invalidation pattern is this PR should work)
- identify the real bottleneck source of this performance issue (which might be an interesting point to solve as it's the real issue)

This PR try to solve the 2 first point together and I don't think this is a good idea.
So, just so I'm clear, I'm not saying that they shouldn't be solved, I like the idea of having proper HTTP cache handling as well as having more efficient applicative cache use. I just think this is something that need to be splitted because it's 2 distinct problems with 2 distinct solutions.